---
title: "Estimating Unique Adult Genotypes with Low-Coverage Whole-Genome-Sequencing"
output:
  html_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(pegas)
library(knitr)
library(igraph)
library(RColorBrewer)
library(tidyverse)
library(cowplot)
library(tinytex)
```

```{r, include = FALSE}
################
#load all our shit
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
#mitochondrial genome fasta of all samples
fasta <- c('data/unamplified_amplified_mf.fas')
#ibsrelate output
genotype_dist = 'data/ibs.model0.results.ibspair'
#source file with ibs functions
source("bin/lib/ibsfunctions.R")
#source file that generates metadata
source("bin/lib/metadata.R")
###
source("bin/lib/ibslib.R")
```

In this report, I present the initial results of a Bill and Melinda Gates Foundation funded project, in partnership with the Drugs for Neglected Diseases Initiative. The overarching aim of the project is to develop a method for identifying whether patients infected with *Onchocerca volvulus*, post-treatment, are infected with worms that survived the treatment initially, or are worms that infected the patient following drug clearance.

Skin-snip biopsies for microfilaria (*mf*) can be used as a proxy for adult sampling - as adults are embedded in tough nodules. As mf are morphologcially indistinguishable from one another, we need to use genetics to identify whether mf from pre- and post- treatment samples are siblings with respect to one another.

The identification of siblings is critical because if the adult female worms (living in their nodules) have survived treatment, then we expect pre- and post- treatment *mf* to be related as full (same mother and father) or half (same mother *or* father) siblings. If the adult female worms have not survived treatment, we do not expect to identify siblings in pre- and post-  treatment mf samples.

Inferring the number of adults that have contributed to the microfilaria pool, and then using the number of unique adult genotypes, to infer the likely number of adults that infect a host. This can then be used to calculate the certainty with which one predicts treatment success or failure.

Here, I present an initial method for estimating sibling relationships and number of breeders from WGS data derived from *O. volvulus* microfilariae. The reconstruction of full-sibling and half-sibling relationships allows one to identify potential treatment failure. The estimation of the number of breeders allows you to then model potential undersampling of mf with respect to the total adult worm burden.

```{r buildmetadata,include = FALSE, echo = FALSE}
#build metadata
metadata <- getseqandsamp()
#do some subs to get simpler sample names
metadata$newsampleid <- gsub('711-004-6-1_g', 'A', metadata$newsampleid)
metadata$newsampleid <- gsub('711-004-6-3_c', 'B', metadata$newsampleid)
metadata$newsampleid <- gsub('010-243-2-1_2_3_III', 'C', metadata$newsampleid)
```

I calculate the IBS statistics from the genotype distribution, calculated by IBSRelate for all pairs of the samples.
The output of interest are the statistics R1, R0 and KING-Robust Kinship. The statistics are described in more detail by **Waples *et al*, 2018** [2]. (**open access**). 

```{r doderived, echo = FALSE}
ibsrelateoutput <-  do_derived_stats(read_ibspair_model0(genotype_dist))
```

Then, I filter out non-mf comparisons and add some more data to the ibsrelate output, per pair of samples:

1. Whether both samples are from the same brood or not
2. The metadata for each samples on either side of the comparison
3. Whether a comparison is between mf - mf or mf - adult
4. The fraction of the total GLs in the callset used to calculate the IBS statistics



```{r relwrangle}
rel <- relmatrixwrangle(ibsrelateoutput, metadata)
rel <- rel %>% filter(typecomp == 'mf_mf')
rel %>% select(sample_id.x, sample_id.y, R0, R1, Kin, typecomp, fracsites, degree_bro) %>% head() %>% kable()
```

**Table 1: A subset of the pairwise relatedness data showing sample ids, IBS coefficients, type of comparison, fraction of sites involved in calculation, and the known degree of relationship - whether a sample is taken from the same brood or not**

Now, I plot the R0/R1 and R1-Kin from the relatedness data. In [[1]], the pattern of clustering in these plots clearly delineates the relationship between individuals. I filtered out comparisons that were missing >20% of the total GL callset.

```{r plotr0r1, fig.height = 3, fig.width = 8, fig.align = "center", echo = FALSE}
#the following functions apply a filter of 0.5 (but mostly the same results up to 0.8)
#################
#Plot R1/R0
r1r0 <- plotr1r0(rel, 0.8)

#################
#Plot R1/R0
r1king <- plotr1king(rel, 0.8)

#################
#Based on these plots, define thresholds for different values of relatedness
#this groups individuals based on values of R0, R1 and KING
rel <- definetherelationship(rel)

######################
#create r0/r1 and r1/king plot

# extract the legend from one of the plots
legend <- get_legend(
  # create some space to the left of the legend
  r1r0 + theme(legend.box.margin = margin(0, 0, 0, 12)) +
    theme(legend.position = "bottom")
)

prow <- plot_grid(
  r1r0 + theme(legend.position="none"),
  r1king + theme(legend.position="none"),
  align = 'vh',
  labels = c("A", "B"),
  hjust = -1,
  nrow = 1
)

plot_grid(prow, legend, ncol = 1, rel_heights = c(1, .1))
```

**Figure 2: Plots of R0/R1 (A) and R0/King-Robust Kinship (B) coefficients for pairs of intrauterine microfilaria. Point colour (Known Degree) indicates whether microfilaria of a pair are from the same brood or not.**


In these two plots one can see clearly defined groupings - lower R0 (<0.6) and higher R1 (>0.5) and Kin (>0.25)groups are are full-sibs (reflected also by [1] Figure 3.) The cluster containing some within-brood pairs between R0 ~ 0.8 and 0.2 looks to me to be half-sib, roughly consistent with [1] also - although apparently deviations from this can depend on population. I have calculated *F* for all the individuals in this sample group (can provide if needed) and they are all highly outbred, so I don't expect inbreeding to be an issue. So, I use the R0, R1 and Kin values to derive a relationship for each pair - full-sibling (FS) or half-sibling (HS)

```{r plotinferredrel}
plotinferredrel(rel)
```

**Figure 3: Plot of R0/R1-Robust-kinship coefficients for pairs of intrauterine microfilaria. Point colour indicates the inferred relationship**

Now, I can plot these relationships using the iGraph library so we can get a better estimate of whether the relationship assignment has worked. We should see three groups of full or half siblings that cluster by brood. One of the broods is taken from an patient in a different village, so I'm not expecting to see any FS/HS relationships between this and the other two, which are taken from the same patient. 


So, here, I pass the relatedness dataframe, and the metadata, and use it to generate edges (links/branches) and vertices (nodes) that can be used as an input for iGraph, gtenerating a colour palette based on the female from which the mf was dissected.

```{r makeigraph}
out <- makeinitialigdfs(rel, metadata)
sibnet <- graph_from_data_frame(out[[1]], out[[2]], directed = FALSE)
palstuff <- make_ig_color_pals_from_samples(out[[2]], "Dark2")
E(sibnet)$color <- as.factor(E(sibnet)$relationship)
```

Now, I plot:

```{r plotig}
{plot(sibnet, vertex.label.cex = .7,  vertex.size=30, vertex.color=palstuff[[3]], vertex.label=NA) 
  legend("topleft",bty = "n",
         legend=levels(palstuff[[2]]),
         fill=palstuff[[1]], border=NA)}
```

**Figure 4: Undirected graph indicating inferred relationships between 10 intrauterine microfilaria. Node label indicates sample name. Node colour indicates worm brood (detailed in legend) (the dam from which it was dissected). Edge colour indicates inferred relationship, where yellow = FS, blue = HS, green = Unrelated** 

Unsurprisingly, we have a mess. The FS clusters are very clear: clustering by yellow lines. However, the HS values are all over the place and complicate things somewhat. We have 10 mf from 3 broods (adult female uteruses). They group into full-sub clusters by brood (e.g. all the individuals from *III* - the gold brood - are FS with respect to once another. Likewise with *63G* - the green brood. *61G* is interesting as it has 3 individuals from the same brood, two of which are FS with respect to each other, with one half-sib. This indicates that we have 1 dam fertilised by two sires).

There's a bit of noise and inconsistency among the HS assignments too:. I think there's a difficulty drawing distinctions between HSs and cousins, so let's drop the HSs for now, and discuss below.

```{r tidyupigraph}
#filter igraph edge dataframe to extract full-sib vertices only
out[[1]] <- out[[1]] %>% filter(., relationship == 'f-sib')
sibnet <- graph_from_data_frame(out[[1]], out[[2]], directed = FALSE)
```
```{r plottidyigraph}
{plot(sibnet, vertex.label.cex = .7,  vertex.size=35, vertex.color=palstuff[[3]],vertex.label=NA)
  legend(-2, 1, legend=c("A", "B", "C"), fill=c("#1b9e77", "#d95f02", "#7570b3"), bty='n',cex=1)
  }
```

**Figure 5: Undirected graph indicating FS relationships (edge colour-orange) amongst 10 intrauterine microfilaria (node label) isolated from three different adult females (node colour).**

This is tidier. We can see clearly FSib clusters coloured by our prior knowledge. 
However, this throws a spanner in the works. We cannot infer the number of total parents from FS clustering alone - we need to know which individuals may share a mother or a father. As we saw above, our HS assignment is all over the place. I can work out half-siblinghood, however, by looking at mitochondrial genotype sharing between individuals. As mitochondria are inherited maternally, finding out mitochondrial genotype in our samples will tell us which individuals share a mother. By process of elimination, we can then infer the number of unique male and female genotypes from our offspring. It also happens that we get great coverage across the mtDNA in each individual, so this makes mtDNA genotyping quite straightforward in our sample set.

```{r makehaplonet}
iumf_fas <- read.dna(fasta, format='fasta') 
#match up metadata to fasta names
netmetadata = metadata[match(rownames(iumf_fas), metadata$sample_id), ]
#change fasta names to form id
names(iumf_fas) <- netmetadata$newsampleid

#make hap object from fasta
onchohaps <- haplotype(iumf_fas)

hnet <- haploNet(onchohaps)

#make table of now many individuals from each form belong to each cluster
ind.hap<-with(
  stack(setNames(attr(onchohaps, "index"), rownames(onchohaps))),
  table(hap=ind, individuals=names(iumf_fas)[values]))


{setHaploNetOptions(pie.colors.function=c("#1B9E77","#D95F02","#7570B3"))
plot(hnet, attr(hnet, "freq"), fast = TRUE, legend = c(-700, 30), show.mutation = 2, scale.ratio = 1, cex = 0.8, labels=FALSE, threshold = 0,  pie=ind.hap)
legend(-30, 6, legend=c("A", "B", "C"), fill=c("#1B9E77","#D95F02","#7570B3"), bty='n', cex=1)}
```

![**Figure 5: Medium spanning network of O.volvulus microfilarial mitochondrial haplotypes from 3 broods of intrauterine mf**](figures/unamplified_amplified_mf_msn.png)



The mitochondrial haplotype network, which I ended up making with the amplified dataset also, (no coverage over nuclear genome though) resolves as three unambiguous clusters, with some diversity within each cluster. However, I have enough information here to clearly define mitochondrial groups.

With more intrauterine mf being sequenced, I will hopefully be able to define mitochondrial groups more robustly. I can define mixing coefficients and assign each mitochondrial sequence to a group with a probability.

Now, I extract the network into an iGraph, so I can parse out information on which sample belongs to which mitochondrial grouping.


```{r getigraphfromhnet, include = FALSE}
ig <- mito_to_haplo_igraph(metadata, fasta, onchonet)
#create the mitochondrial igraph ibject
mitonet <- graph_from_data_frame(ig$edges, ig$vertices, directed = FALSE)
#generate vertext colours for the mitochondrial igraph
mitopal <- make_ig_color_pals_from_samples(ig$vertices, "Dark2")
```

And then plot the mitochondrial groups as a graph, setting a cutoff of >3 SNP distances between mitochondrial to no longer being in the same cluster.

```{r clusters, echo = FALSE}
fsibgrups<-fastgreedy.community(sibnet)
mitogrups<-fastgreedy.community(mitonet)

#plot mitochondrial clustering
{plot(mitogrups, mitonet, vertex.label.cex = .7,  vertex.size=30, col=mitopal[[3]])
  legend("topleft",bty = "n",
         legend=levels(mitopal[[2]]),
         fill=mitopal[[1]], border=NA)}
```

**Figure 6: Graph of mitochondrial haplotype relationships -  nodes indicate mitochondrial haplotypes, vertices indicate pairwise distance of <3 SNPs. Polygons indicate defined clusters**

**Figure 6** is a graph of mitochondrial haplotypes that corresponds to the mitochondrial network in **Figure 5**. As in **Figure 5**, the haplotypes group according to the female from which they were extracted. Each haplotype is connected to other members of the group by no more than 3 SNPs - the cutoff for determining a group or not. (I can also use an mtDNA phylogeny for this but the network was easier to match in iGraph). A simple greedy clustering algorithm collects the mitochondrial haplotypes into their groups and assigns an identifier to them.

```{r plotfsclustering, echo = FALSE}
#plot fullsib clustering
{plot(fsibgrups, sibnet, vertex.label.cex = .7,  vertex.size=30, col=palstuff[[3]],vertex.label=NA) 
  legend("topleft",bty = "n",
         legend=levels(palstuff[[2]]),
         fill=palstuff[[1]], border=NA)}
```

**Figure 7: Graph of sibling relationships -  nodes indicate individual mf coloured by known brood. Polygons indicate FS clusters**

**Figure 7** is a graph of individual mf linked by FS relationships, and the clusters I have defined for them. 

Now, we need to compare the FS and mitochondrial clusters, and see if there are any non-FS mf that are part of the same mitochondrial group - and are therefore maternal half-sibs. I will collapse the FS clusters into single vertices, and then plot membership of the same mitochondrial group as edges: if we see FS groups linked by mitochondrial edges then we know that they are maternal half-siblings (as they are not FS but are members of the same mitochondrial group).

```{r makemaps, echo = FALSE}
ind.haps <- getindhap(fasta)
mitomap<-makemitomap(mitonet, mitogrups, ind.haps)
fsibmap<-makesibmap(sibnet, mitomap, fsibgrups)
```


```{r makefinalig,  echo = FALSE}
mitosibedge <- rel %>% 
  left_join(., mitomap, by = c("sample_id.x" = "sample_id")) %>% 
  left_join(., mitomap, by = c("sample_id.y" = "sample_id")) %>% 
  mutate(mitosib = ifelse(mitogroup.x == mitogroup.y, 'mitosib', 'mitononsib')) %>% 
  select(sample_id.x, sample_id.y, relationship, mitosib) %>% 
  drop_na() %>% 
  left_join(., fsibmap, by = c("sample_id.x" = "sample_id")) %>% 
  left_join(., fsibmap, by = c('sample_id.y' = 'sample_id')) %>% 
  select(., fsibgroup.x, fsibgroup.y, mitosib) %>% 
  filter(., mitosib == 'mitosib') %>% 
  filter(fsibgroup.x != fsibgroup.y)

#create vertices from membership dataframe
mitosibvertex <- fsibmap %>% 
  select(fsibgroup, newsampleid) %>% distinct()

#create graph object, fill by palette we created above, with corresponding legend
msibgraph  <- graph_from_data_frame(mitosibedge, mitosibvertex, directed = F)
msibgraph <- igraph::simplify(msibgraph, remove.multiple=TRUE)

mitopal <- make_ig_color_pals_from_samples(mitosibvertex, "Dark2")
```

```{r plotfinalig, echo = FALSE}
{plot(msibgraph, vertex.label.cex = .7,  vertex.size=40, vertex.color=mitopal[[3]],vertex.label=NA)}
```

**Figure 9: Graph indicating fs and mitochondrial relationships of intrauterine mf. Nodes indicate FS clusters, edges indicated shared mitochondrial haplotypes**

In **Figure 9** we observe the four FS groups, with two FS groups (from 61G - blue) belonging to the same mitochondrial group - they are connected by a grey edge indicating sharing of a mitochondrial group. To calculate the number of parents contributing to this pool of individuals, we suppose 2 parents per FS group, subtracting one parent for each distinct mitochondrial relationship (so, where two FS groups actually share a parent - so 1 mother and 2 fathers).

The number of unique parents in our pool of individuals: (2*no.vertices - no. edges):
```{r uniqueparents, echo = FALSE}
mitodf <- igraph::as_data_frame(msibgraph, "both")
count(mitodf$vertices)*2 - count(mitodf$edges)
```
The number of unique mothers: (no. vertices in the mitochondrial graph):
```{r uniquemothers, echo= FALSE}
count(mitodf$vertices) - count(mitodf$edges)
```

[1]: https://link.springer.com/article/10.1186/s13071-019-3687-1
[2]: https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14954











