---
title: "Estimating Unique Adult Genotypes with Low-Coverage Whole-Genome-Sequencing"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(pegas)
library(knitr)
library(igraph)
library(RColorBrewer)
library(tidyverse)
```


```{r}
################
#load all our shit
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
#mitochondrial genome fasta of all samples
input <- c('/Users/tristanpwdennis/Projects/onchogenome/data')
#ibsrelate output
testpath0 = '/Users/tristanpwdennis/Software/onchotoolbox/data/ibs.model0.results.ibspair'
#source file with ibs functions
source("ibsfunctions.R")
#source file that generates metadata
source("metadata.R")
###
source("~/Projects/onchogenome/scripts/ibslib.R")
```

```{r}

```























